base/range.jl

```diff
@@ -41,7 +41,7 @@ Range operator. `a:b` constructs a range from `a` to `b` with a step size of 1 (
 # without the second method above, the first method above is ambiguous with
 # (:)(start::A, step, stop::C) where {A<:Real,C<:Real}
 function _colon(start::T, step, stop::T) where T
-    T′ = typeof(start+step)
+    T′ = typeof(start+zero(step))
```

Comment:
```
function _colon(start::T, step::TS, stop::T) where {T,TS}
    T′ = promote_type(T,TS)
...
```

Be more sensible?

Though, if one is defining addition, to not use type-promotion,  my suggestion wouldn't work.
But that is a pathological case, and at that point are we even sure `zero` will be defined?

Reply:
I have thought about it. The problem is that when you take `'a':1:'c'` then:
```
julia> promote_type(Char, Int)
Any
```
so I have thought that assuming `TS` defines `zero` is more sensible (typical steps like numbers or `Day` or `Millisecond` define `zero`). Not 100% sure though.

Reply:
Ah, fair enough.